danascheider/skyrim_inventory_management_frontend
# Enable user to create new shopping lists from the UI

## Context

[**Enable user to create new shopping lists from the UI**](https://trello.com/c/uBb2dyec/8-enable-user-to-create-new-shopping-lists-from-the-ui)

So far, users can edit shopping list titles from their dashboard but can't create or remove the lists, nor can they add list items.

## Changes

* Add creation form to the `ShoppingListPage`
* Change the `ShoppingListForm` to `ShoppingListEditForm` since `ShoppingListCreateForm` is different
* Add `performShoppingListCreate` function to the `ShoppingListProvider` enabling the form to create the shopping list
* Tweak handling of auth errors from the API
* Make some changes to the `ColorContext` and how it is used to fix some bugginess
* Make Storybook work

## Manual Test Cases

Users should be able to create new lists through the UI, provided the title they give is valid (contains only alphanumeric characters and spaces, isn't "Master") and hasn't been used for another list.

Make sure users can create new lists:

* When they have existing lists
* When they have no lists
* When they have just created a list and haven't refreshed the page yet

Newly created lists should be added directly below the master list, before any of the other regular lists.

If a user doesn't have any lists, a master list will also be created and be returned from the API, so the page the user sees should always either have no shopping lists or 2+.

## Screenshots and GIFs

I just noticed that giphy capture is messing up the colours in this. The flicker in that "Create" button is not actually there. The button is blue.

### Creating Successfully When No Lists Exist

![issue-8-1](https://user-images.githubusercontent.com/5115928/123881578-5c2a7c80-d988-11eb-9bc7-854fa486243e.gif)

### Creating Unsuccessfully When No Lists Exist

![issue-8-4](https://user-images.githubusercontent.com/5115928/123882133-7e70ca00-d989-11eb-8064-3edae933279d.gif)

### Creating Successfully When Other Lists Exist

![issue-8-2](https://user-images.githubusercontent.com/5115928/123881600-69e00200-d988-11eb-9e8c-987c7f3d51ea.gif)

### Creating Unsuccessfully When Other Lists Exist

![issue-8-3](https://user-images.githubusercontent.com/5115928/123881633-76645a80-d988-11eb-886f-24c86d8d4fef.gif)

### Form Disabled When Loading State is "error"

<img width="1430" alt="Screen Shot 2021-06-30 at 1 45 46 pm" src="https://user-images.githubusercontent.com/5115928/123898963-a708bc00-d9a9-11eb-9bd6-3661c34c66cb.png">

### Form Disabled When Loading State is "loading"

<img width="1432" alt="Screen Shot 2021-06-30 at 1 46 00 pm" src="https://user-images.githubusercontent.com/5115928/123898983-b0922400-d9a9-11eb-9c29-65932c323140.png">

## Note on Storybook

In the Storybook story for `ShoppingListPage` called "Empty", if you add an item to the list, it and the master list get created as expected, however, if you create further lists they'll just replace the ones you made. That's because MSW is configured to mock the request to return two elements (the master list and the list the user requested). It's a bit of a project to get it to respond differently to subsequent requests and I don't think it's worth the trouble - we'd basically be making a whole mock server and I don't want to deal with that. Not to mention, mock servers are known to cause problems when they deviate from behaviour of the actual server - which is exactly what this already does. So I'm hesitant to dig myself in deeper.
